2.1.0

KillerOfPie opened the pull request:
[contributing]: https://github.com/SparklingComet/TradeShop/blob/master/.github/CONTRIBUTING.md


# Ready To be checked

I may have built this on my 2.0.1 and had to push to 2.1.0 once done hopefully everything from 2.0.1 got through

# Release Notes

https://github.com/SparklingComet/TradeShop/wiki/2.1.0-Release-notes

KillerOfPie committed:
Fixed bugs and missing features
- Fixed 1.14 support(hopefully didnt actually test anything in 1.14
- Added JoinEventListener and ShopProtectionListener
- Changed api version back to 1.13 so that it should work on both 1.13 and 1.14
- Re-implemented Explosion protection
- Added hopper blockage on shop chests

KillerOfPie committed:
Fixed Bugs and added Features
+ Added MULTI command
+ Added Plugin_Behind message to config(message sent to players with admin perm when they log in)
+ Added Multi_Update and Multi_Amount messages for new Multi command
+ Added Allow_Multi_Trade setting
+ Added per player setting reset to join event
+ Added Admin Update message to join event
+ Added Event param to Shop for temporary storage to fix Sign update not running during event and causing errors
+ Changed Shop#missingItems to Shop#isMissingItems
+ Made BukkitVersion#(GetMajor, getMinor, getPatch) to public for version checking onJoin
+ Made JsonConfiguration work with both Shops and per player files(player methods wont work if Json was created using shop method and vice-versa)
+ Utils#containsAtLeast now is passed amount for more accurate checking
+ Added TradeShop#getUpdater for notifications
+ Updated plugin.yml and pom.xml to 2.1.0
- Removed Allow_Double_Trade and Allow_Quad_Trade

KillerOfPie committed:
Fixed Bugs and added Features
+ Fixed Tab completion for Materials not narrowing
+ Added ALLOW_HOPPER_EXPORT setting so server can disable/enable hopper protection per sign type
+ Fixed ShopChest not resetting names
+  Added fixed gameMats name list to ListManager for name tab completion(so that we dont have to load the list everytime some tries to tab complete that)
+ Added ShopChest#getShop(unused for now, no planned usages)

smashedbotatos commented:
Changed to spigot-api 1.14, and built. Working well, but errors on shop breaking. Hoppers no longer take items from cshops, but they will ADD items to a shop.

On breaking a shop this is the error. 
`[22:29:41] [Server thread/ERROR]: Could not pass event BlockBreakEvent to TradeShop v2.1.0
org.bukkit.event.EventException: null
        at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:320) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at org.bukkit.plugin.SimplePluginManager.fireEvent(SimplePluginManager.java:520) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:505) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.PlayerInteractManager.breakBlock(PlayerInteractManager.java:286) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.PlayerInteractManager.a(PlayerInteractManager.java:219) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.PlayerConnection.a(PlayerConnection.java:1203) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.PacketPlayInBlockDig.a(SourceFile:40) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.PacketPlayInBlockDig.a(SourceFile:10) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.PlayerConnectionUtils.lambda$0(PlayerConnectionUtils.java:13) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.TickTask.run(SourceFile:18) [minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.IAsyncTaskHandler.executeTask(SourceFile:135) [minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.IAsyncTaskHandlerReentrant.executeTask(SourceFile:23) [minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.IAsyncTaskHandler.executeNext(SourceFile:114) [minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.MinecraftServer.aW(MinecraftServer.java:895) [minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.MinecraftServer.executeNext(MinecraftServer.java:888) [minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.IAsyncTaskHandler.awaitTasks(SourceFile:123) [minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.MinecraftServer.sleepForTick(MinecraftServer.java:872) [minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at net.minecraft.server.v1_14_R1.MinecraftServer.run(MinecraftServer.java:805) [minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
Caused by: java.lang.NullPointerException
        at org.shanerx.tradeshop.objects.ShopChest.resetName(ShopChest.java:137) ~[?:?]
        at org.shanerx.tradeshop.listeners.AdminEventListener.onBlockBreak(AdminEventListener.java:82) ~[?:?]
        at sun.reflect.GeneratedMethodAccessor213.invoke(Unknown Source) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_212]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_212]
        at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:316) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c]
        ... 19 more`

KillerOfPie committed:
Fixed Bugs
+ Changed unused MISSING_INFO message to MISSING_CHEST
+ Fixed error when shop sign/chest deleted in wrong order
+ Storage on shops can be replaced after being removed(Shop closes as soon as its removed)

KillerOfPie commented:
> Changed to spigot-api 1.14, and built. Working well, but errors on shop breaking. Hoppers no longer take items from cshops, but they will ADD items to a shop.
> 
> After the shop chests are broken, the json is not being deleted. I think it is from this error on breaking the shop.
> 
> On breaking a shop this is the error.
> `[22:29:41] [Server thread/ERROR]: Could not pass event BlockBreakEvent to TradeShop v2.1.0 org.bukkit.event.EventException: null at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:320) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c] at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c] at org.bukkit.plugin.SimplePluginManager.fireEvent(SimplePluginManager.java:520) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c] at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:505) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.PlayerInteractManager.breakBlock(PlayerInteractManager.java:286) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.PlayerInteractManager.a(PlayerInteractManager.java:219) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.PlayerConnection.a(PlayerConnection.java:1203) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.PacketPlayInBlockDig.a(SourceFile:40) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.PacketPlayInBlockDig.a(SourceFile:10) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.PlayerConnectionUtils.lambda$0(PlayerConnectionUtils.java:13) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.TickTask.run(SourceFile:18) [minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.IAsyncTaskHandler.executeTask(SourceFile:135) [minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.IAsyncTaskHandlerReentrant.executeTask(SourceFile:23) [minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.IAsyncTaskHandler.executeNext(SourceFile:114) [minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.MinecraftServer.aW(MinecraftServer.java:895) [minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.MinecraftServer.executeNext(MinecraftServer.java:888) [minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.IAsyncTaskHandler.awaitTasks(SourceFile:123) [minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.MinecraftServer.sleepForTick(MinecraftServer.java:872) [minecraft_server.jar:git-Spigot-093165d-cb3d14c] at net.minecraft.server.v1_14_R1.MinecraftServer.run(MinecraftServer.java:805) [minecraft_server.jar:git-Spigot-093165d-cb3d14c] at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212] Caused by: java.lang.NullPointerException at org.shanerx.tradeshop.objects.ShopChest.resetName(ShopChest.java:137) ~[?:?] at org.shanerx.tradeshop.listeners.AdminEventListener.onBlockBreak(AdminEventListener.java:82) ~[?:?] at sun.reflect.GeneratedMethodAccessor213.invoke(Unknown Source) ~[?:?] at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_212] at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_212] at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:316) ~[minecraft_server.jar:git-Spigot-093165d-cb3d14c] ... 19 more`

Literally just fixed all that :P
I left the hopper input working as i figure that's not necessarily a problem that we need to manage, plus allows players to have a little extra storage for their shops.

And shop json isn't deleted when chest is broken, chest is just removed from the shop and the shop is closed(now) thanks for the input though would've been very useful if i hadn't happened to already be working on it.

KillerOfPie commented:
Also @smashedbotatos I made a discord for the plugin help and such if you wanted to join it. https://discord.gg/qzdSUnw

KillerOfPie committed:
Added Features
+ Shops now automatically close when they don't have enough product for a normal transaction
+ Moved Shop#saveShop call into Shop#(setClosed and setOpen) so we dont have to save after every open/close call

KillerOfPie committed:
Added Features
+ Implemented max shops per chunk, default is set to 128(no clue if that's too many or not)

KillerOfPie committed:
Added Features and fixing bugs?
+ Added support for multiple items per side of the trade
+ Added set(Cost/Product) add(cost/product) remove(cost/product) and list(cost/product) commands
+ Small optimizations
+ Small default message changes
+ removed explode setting from ShopType enums
+ Added data converter for versions before this
+ Added new #checkInventory and #canExchangeAll methods to Utils

KillerOfPie committed:
Bug fixes
+ Fix for over sized stacks not being fixed when loading from old file
+ Fix for signs not being updated after being reloaded if changes were made
+ Fix for NPE while in custom inventories
+ Some cleanup for excessive code

KillerOfPie committed:
Bug fixes and New features
+ Added max stack per trade side support(default: 6)
+ Fixed NPE from AdminEventListener
+ Fixed NPE from ShopTradeListener
+ Fixed NPE while creating shops

KillerOfPie commented:
@SparklingComet Im probably going to added one or two more small features but this is basically done if you could give it a look over

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/CommandCaller.java:
```diff
@@ -133,18 +133,36 @@ public boolean onCommand(CommandSender sender, Command cmd, String label, String
+				break;
+			case REMOVE_COST:
+				cmdRnnr.removeCost();
+				break;
```
Indentation is misleading, please fix.

KillerOfPie replied:
Ill fix formatting once everything else is fixed otherwise github shows almost everything as being changed and its hard to find cahnges

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java:
```diff
@@ -153,6 +154,188 @@ public void reload() {
+				sb.append(itm.getItemMeta().getDisplayName());
+			else
+				sb.append(itm.getType().toString());
+			sb.append("\n");
```
Lines 170-178 can be written as:
`sb.append(String.format("&b[&f%d&b]####&2- &f%s\n", counter, itm.hasMeta() && itm.getItemMeta().hasDisplayName() ? itm.getItemMeta().getDisplayName() : itm.getType().toString()));`
(Replace #### with four spaces - github wouldn't let me put four spaces in a row.)

KillerOfPie replied:
Changed

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java:
```diff
@@ -153,6 +154,188 @@ public void reload() {
+				sb.append(itm.getItemMeta().getDisplayName());
+			else
+				sb.append(itm.getType().toString());
+			sb.append("\n");
```
see my previous comment concerning one liners

KillerOfPie replied:
Changed

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/CommandCaller.java:
```diff
@@ -168,7 +189,7 @@ public boolean checkPerm() {
      */
     public boolean checkPerm(TradeCommand exec) {
-        if (!cmdPass.getSender().hasPermission(exec.getPermission()) && !exec.getPermission().equals(Permissions.NONE)) {
+		if (!cmdPass.getSender().hasPermission(exec.getPermission()) && !exec.getPermission().equals(Permissions.NONE.getValue())) {
```
what is this?

KillerOfPie replied:
A method for checking if the player has a perm using a more DRY friendly way, seeing as this same code was being run 5-10 times through the file, and none has to be checked for separately

SparklingComet replied:
I am aware of that, but the .getValue
 part confuses me

KillerOfPie replied:
because that is just checking if the permission enum that is being checked is NONE, because checking for the perm "" never returns true unless they literally have a blank perm

SparklingComet replied:
But why is a permission check for NONE required? It literally means that no permission is required to perform that action

KillerOfPie replied:
Because the command enum lists Permissions.NONE when it doesnt need one otherwise that field would return null and i would still have to check for null, Permissions.NONE is just a more readable way to do it

KillerOfPie replied:
This was messed up a bit, the internal command check should use the enum, however the external check is just a string so thats been changed to check for "none"

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java:
```diff
@@ -153,6 +154,188 @@ public void reload() {
+		Material mat = null;
+
+		if (command.hasArgAt(1) && isInt(command.getArgAt(1))) {
+			amount = Integer.parseInt(command.getArgAt(1));
```
what happens if arg(1) is not a valid int? it this being checked somewhere else or will the method throw a NumberFormatException?

KillerOfPie replied:
#isInt checks if it is an int and returns a boolean

SparklingComet replied:
I skipped that part while reading, sorry

KillerOfPie replied:
Its cool, it hides pretty well

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java:
```diff
@@ -209,7 +392,78 @@ public void addProduct() {
+		Material mat = null;
+
+		if (command.hasArgAt(1) && isInt(command.getArgAt(1))) {
+			amount = Integer.parseInt(command.getArgAt(1));
```
same questions as above.

KillerOfPie replied:
same answer as above, resolving this one

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/CommandTabCompleter.java:
```diff
@@ -158,11 +106,14 @@ public CommandTabCompleter(TradeShop instance, CommandPass command) {
+		for (String str : plugin.getListManager().getGameMats()) {
+			if (str.toLowerCase().contains(request.toLowerCase())) {
+				toReturn.add(str);
+			}
```
what is this about?

KillerOfPie replied:
It returns a list of all material names that partially match what the player has type so for the add/set product/cost commands they can tab complete material names and the more they type the more itll narrow down available types

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enumys/Commands.java:
```diff
@@ -42,17 +42,24 @@
+	REMOVE_COST(Lists.newArrayList("removeCost", "delCost"), Permissions.NONE, 2, 2, true, "Removes a product from the shop", "/tradeshop $cmd$ <List #>"),
+	LIST_PRODUCT(Lists.newArrayList("listProduct"), Permissions.NONE, 1, 1, true, "Lists the products in the shop", "/tradeshop $cmd$"),
+	LIST_COST(Lists.newArrayList("listCost"), Permissions.NONE, 1, 1, true, "Lists the costs in a shop", "/tradeshop $cmd$"),
+	OPEN(Lists.newArrayList("open"), Permissions.NONE, 1, 1, true, "Open shop", "/tradeshop $cmd$"),
```
quick indentation fix wouldn't hurt

KillerOfPie replied:
same as top, ill do it once everything else is done

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enumys/Message.java:
```diff
@@ -145,7 +155,13 @@ public static void setDefaults() {
+		addMessage("multi-amount", "&aYour trade multiplier is %amount%.");
+		addMessage("too-many-chests", "&cThere are too many shops in this chunk, you can not add another one.");
+		addMessage("shop-item-list", "&aThe shops %type%:\n%list%");
+		addMessage("too-many-items", "&cThis trade can not take any more %side%!");
```
nice addition.

KillerOfPie replied:
Thanks

SparklingComet commented on src/main/java/org/shanerx/tradeshop/enumys/ShopStatus.java:
```diff
@@ -31,7 +33,7 @@
 
 	private String colour;
-	private static final char COLOUR_CHAR = '&';
+	private transient static final char COLOUR_CHAR = '&';
```
Gson should not be serializing static fields...

KillerOfPie replied:
I know i was just being paranoid, because it also shouldnt serialize parent classes that dont extend serializable but it was trying to serialize utils. changed

SparklingComet replied:
I think it might be indeed supposed to serialize all fields of an object unless instructed otherwise. Inherited objects fully belong to the extending class, which is probably why Gson was attempting to serialize the Plugin instance.

KillerOfPie replied:
I saw it somewhere but i cant find it now super classes shouldn't be serialized unless they also extend serializable, but that doesn't matter anymore since its not extending it

Commit:
Reformatted CommandCaller

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopProtectionListener.java:
```diff
@@ -0,0 +1,117 @@
+			if (plugin.getListManager().isInventory(b.getType()) &&
+					((Nameable) b.getState()).getCustomName() != null &&
+					((Nameable) b.getState()).getCustomName().contains("$ ^Sign:l_")) {
+				if (Shop.loadShop((new ShopChest(b.getLocation())).getShopSign()) != null && !Setting.findSetting(Shop.loadShop((new ShopChest(b.getLocation())).getShopSign()).getShopType() + "SHOP_EXPLODE".toUpperCase()).getBoolean())
```
This line is way longer than it should be.
Why are you concatenating the shop type with “SHOP_EXPLODE”?
And why are you calling “SHOP_EXPLODE”.toUpperCase() when the string is already uppercase?

KillerOfPie replied:
'shop.getShopType() + "SHOP_EXPLODE".toUpperCase())'
because i want to make sure that the whole string is uppercase, even though shoptype is an enum and should be uppercase

also fixed, rewrote that logic section to not be as excessive, it was loading the shop like 3 or 4 times

SparklingComet replied:
But that is not what was happening, you were only invoking the method on the latter part

KillerOfPie replied:
you right, fixed it

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopTradeListener.java:
```diff
@@ -55,189 +60,222 @@ public void onBlockInteract(PlayerInteractEvent e) {
 
-		if (e.getAction() == Action.RIGHT_CLICK_BLOCK) {
+		if (ShopType.isShop(e.getClickedBlock())) {
+			s = (Sign) e.getClickedBlock().getState();
```
And what happens when #getClickedBlock is null?
PlayerActionEvent gets triggered when Players hit air too

KillerOfPie replied:
'This event will fire as cancelled if the vanilla behavior is to do nothing (e.g interacting with air).'
- https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/event/player/PlayerInteractEvent.html

It shouldnt trigger if the block is air

Also ShopType#isShop will return false if the block being checked is not a type of sign

SparklingComet replied:
Fair enough, missed that part too.
However, the event is indeed triggered when air is hit:

> Represents an event that is called when a player interacts with an object *or air*, potentially fired once for each hand. The hand can be determined using getHand().

KillerOfPie replied:
yea but it should cancel before it hits us anyways

SparklingComet replied:
You are right, though even if it is canceled, once Bukkit fires the event all listeners get executed.
the isCancelled check should block those cases, though

KillerOfPie replied:
it shoud, hoowever like i said isShop already checks for for material so an additional check isnt necessary

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopTradeListener.java:
```diff
@@ -55,189 +60,222 @@ public void onBlockInteract(PlayerInteractEvent e) {
 
-			JsonConfiguration json = new JsonConfiguration(s.getChunk());
-			shop = json.loadShop(new ShopLocation(s.getLocation()));
+		if (shop == null)
```
Is this even possible? You checked whether the clicked block is a shop

KillerOfPie replied:
Just because the sign is a shop sign doesn't mean that there is still a shop saved in file, this is to make sure the shop was actually loaded otherwise it throws NPEs

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/Shop.java:
```diff
@@ -448,19 +563,28 @@ public boolean hasCost() {
+			saveShop();
+			updateSign();
+			return true;
+		} catch (IndexOutOfBoundsException ex) {
```
Why not check for product.size at the beginning? No point throwing an exception if you can avoid it

KillerOfPie replied:
because im stupid sometimes, changed

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/ShopChest.java:
```diff
@@ -128,24 +131,25 @@ public String getName() {
+							((Nameable) getOtherHalfOfDoubleChest(chest).getState()).getCustomName().split(sectionSeparator)[0]);
+				}
+
+				bs.update();
```
This goes for any part where you are renaming double chests.
‘bs’ is the BlockState object for the first half. Since both halfs are being renamed, shouldn’t the BlockState of the second half be updated too?

KillerOfPie replied:
probably, changed

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/Utils.java:
```diff
@@ -52,13 +59,15 @@
+	protected PluginDescriptionFile pdf = plugin.getDescription();
+	protected String PREFIX = "&a[&eTradeShop&a] ";
+	private UUID KOPUUID = UUID.fromString("daf79be7-bc1d-47d3-9896-f97b8d4cea7d");
+	private UUID LORIUUID = UUID.fromString("e296bc43-2972-4111-9843-48fc32302fd4");
```
All uppercade fields here should be final

KillerOfPie replied:
I changed how they were being set when i was trying to get around the Gson error and forget to set them back, fixed

SparklingComet reviewed (dismissed):
Looks good, but there are a couple oddities that might sprout bugs along the way

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java:
```diff
@@ -549,6 +861,31 @@ public void addMember() {
+			int amount = 2;
+
+			if (isInt(command.getArgAt(1)))
+				amount = Integer.parseInt(command.getArgAt(1));
```
Again ^

KillerOfPie committed:
Bug fixes
+ Fixed poorly written logic in a few places
+ Shortened excessive append calls in CommmandRunner
+ Removed unnecessary transient
+ Replaced missing final
+ Removed unneeded try/catch
+ Added missing blockstate updates

KillerOfPie committed:
Merge remote-tracking branch 'origin/2.1.0' into 2.1.0

KillerOfPie committed:
Bug fixes
+ Fix for bad #toUpperCase calls

KillerOfPie committed:
Bug fixes
+ Fix bad perm checks
+ Made debug messages for signs a little more 'useful'
+ removed unnecessary null check
+ updated pom for bukkit 1.14.2

KillerOfPie committed:
Bug fixes
+ small perm changes

KillerOfPie committed:
**Code Reformat**

KillerOfPie committed:
Minor changes:
+ Removed windowShop alias for what

KillerOfPie committed:
Bug fixes:
+ Fixed bug where shops could be made without permissions
+ Added checking for admins when opening/closing shops(both for iTrade and normal shops)
+ Possibly fixed bug were items were not properly traded after filling a stack
+ Fixed bug with player who created iTrade being named in the trade instead of the Server Shop name from config
+ Fixed bug where iTrade would not create because it was missing a chest

KillerOfPie committed:
Bug fixes:
+ Fixed bug with improper amounts being traded
+ Also some DRY improvements
+ Updated POM for 1.14.3

KillerOfPie committed:
Bug fixes:
+ Possible fix for plugin not properly working on initial load

KillerOfPie committed:
Bug fixes:
+ Fixed Explosion protection not working in 1.14
+ Added Version check so Explosion protection should still work in 1.13

KillerOfPie committed:
Update README.md

KillerOfPie committed:
Changed Version to 2.1.0-DEV until its ready for release

KillerOfPie committed:
Merge remote-tracking branch 'origin/2.1.0' into 2.1.0-newer

KillerOfPie committed:
Ready for Release, Version set to 2.1.0

KillerOfPie commented:
I've already added all the features that I want in this version, so its ready. In 2.2.0 we need to have the Add-On Support done so i can work on the add-on plugins for Economy and Custom Items.

KillerOfPie committed:
Fix for no perm message during completion before command being run.

Commit:
Bug Fixes
- Fix for 'what' command showing players name and not server shop name
- Fix for 'trade' and 'itrade' shops using 'bitrade' trading

Commit:
Bug Fixes and Maven Update and Clean-up
- Fix for Server Shops not being breakable (changed where TradeListener was cancelling events)
- Moved all Listeners from EventListener to ProtectionListener
- Removed a bunch of random debug code
- Updated Maven Shade version to 3.1.0
- Updated bstats Repo to codemc
- Updated spigot and bukkit to  1.14.4R0.1
- Updated bstats dependency to 1.5

SparklingComet reviewed (commented):
Was this both tested on 1.13.x and 1.14.y?

SparklingComet committed:
Merge branch 'master' into 2.1.0

SparklingComet commented:
^^ fixed merge conflicts

KillerOfPie commented:
Because it is works on both versions
